infrastructure/data_access/http/synthetics_repo.py

```diff
@@ -56,9 +62,35 @@ def _get_mesh_test_rows(self, test_id: TestID, results_lookback_seconds: int) ->
         return transform_to_internal_mesh_rows(most_recent_result)
 
     def _get_agents(self, test_id) -> Agents:
+        agent_ids = self._get_agent_ids(test_id)
+
+        with self._cached_agents_lock:
+            self._update_agents_cache_if_needed(agent_ids)
+            return self._cached_agents.filter(agent_ids)
+
+    def _get_agent_ids(self, test_id):
+        logger.debug(f"Getting agent IDs for test with ID {test_id}")
         test_resp = self._api_client.synthetics_admin_service.test_get(test_id)
+        agent_ids = test_resp.test.settings.agent_ids
+        logger.debug(f"Got agent IDs for test with ID {test_id}: {agent_ids}")
+        return agent_ids
+
+    def _update_agents_cache_if_needed(self, agent_ids):
```

Comment:
Let's keep it simple (assuming that `Agents.list_agent_ids` returns `set`, as I've suggested in my other comment):

```
def _update_agents_cache_if_needed(self, agent_ids):
    if self._cached_agents.list_agent_ids() != set(agent_ids):
       self._update_agents_cache()
```
and get rid of the `_is_cached_agents_update_necessary` method. BTW, this also handles the case of an agent stored in the cache and no longer used in the mesh test. 

BTW, properties of an agent (like IP address, name and alias) can (in theory) change without the ID changing (e.g. there is PATCH /synthetics/v202101beta1/agents/{agent.id} API method). I don't think it is likely to happen, but let's update agent cache periodically (period should be configurable, default say 60 minutes) to stay on the safe side.